pomodoro-client.py

```diff
@@ -76,47 +79,77 @@ def main():
 @click.command()
 def status():
     pomodoro = get_pomodoro_proxy()
-    pomodoro_data = extract_pomodoro_data(pomodoro)
-    click.echo(format_output(pomodoro_data))
+    if pomodoro is not None:
```

Comment:
Your MR looks very good to me, thanks for the contrib :smile:!

Just one note. I can see that this check created a very repetitive pattern in the code.

It always had the line 

```python
pomodoro = get_pomodoro_proxy()
```

but now it has

```python
pomodoro = get_pomodoro_proxy()
    if pomodoro is not None:
        # stuff
    else:
        click.echo("gnome-pomodoro is not running")
```

I am not actively working with Python code anymore, but if I remember correctly, the decorator pattern  is the best way to get rid of this duplication. I imagine something like this:

```python
@click.command()
def command_with_pomodoro_proxy(unwrapped_function):
    def wrapped_function():
        pomodoro = get_pomodoro_proxy()
        if pomodoro is not None:
            unwrapped_function(pomodoro)
        else:
            click.echo("gnome-pomodoro is not running")
```

then, all the commands could be defined in a simpler way, something like this:


```python
command_with_pomodoro_proxy
def pause(pomodoro):
    pomodoro.Pause()
```

as it's been a long time since I touched Python code, I'm not 100% confident that this code is correct but I think you get the idea :-)

Anyway, I like the pull request as is, but I'd be really grateful if you could also include that refactor (or something similar, if you have a better suggestion)

Reply:
Sure, I can take a look into using decorators here! :)
I'll study them, play around with them a bit, then modify my PR to incorporate decorators.

Reply:
Thanks! Feel free to ask for help if you're stuck!

Reply:
I've updated the PR to use decorators as you've suggested.
Let me know what you think. :)

Reply:
Thanks for updating with decorators. Sorry for being such a difficult reviewer, but I still have a suggestion.

While now you are using decorators and the actual logic is not duplicated, you also introduced a lot of boilerplate, so in the end this decorator implementation didn't solve the redundancy issue.

I see now that there was a problem with my original proposal, as click takes function names for command names, so click cannot be applied inside the operators.

However, decorators can be chained, so there's no reason why we couldn't apply 2 decorators to the same function, i. e. click doesn't have to be applied inside the decorator.

Therefore, this should be possible:

```python
@click.command()
@with_pomodoro_proxy
def reset(pomodoro_proxy):
    pomodoro_proxy.Reset()
```

That way only 1 extra line is added to each command.

The way to implement that would be something like this:

```python
from functools import wraps

def with_pomodoro_proxy(f):
    @wraps(f)
    def wrapped_command():
        pomodoro = get_pomodoro_proxy()
        if pomodoro is not None:
            f(pomodoro)
        else:
            click.echo("gnome-pomodoro is not running")
    return wrapped_command
```

The reason why `functools.wraps` is required is because manually wrapping will also change the function name. `functools.wraps` will wrap the function with the proxy getter functionality without changing it's name, hence it should work fine with click.

Note that the definition for the inner command takes the proxy as an argument:

```
def reset(pomodoro_proxy):
    pomodoro_proxy.Reset()
```

once you add our new decorator (if it runs at first try  that is), then it should effectively turn it into a function with zero arguments:

```python
@with_pomodoro_proxy
def reset(pomodoro_proxy):
    pomodoro_proxy.Reset()
```

which is because of the signature of the wrapped command inside the decorator:

```
    def wrapped_command():
```.

therefore it should now work fine with click with a much easier syntax, only at the cost of making the decorator itself a little bit more complicated

Reply:
> Sorry for being such a difficult reviewer,
Don't worry about that. :)  Your feedback is very helpful.  I'm not a Python expert, I usually use Python to write quickie scripts for my machine.  So I'm learning quite a bit from your feedback! :)
I'm off to dinner, but I'll take a look at this idea afterwards.  Thanks! :)